tests/trainer/test_data_collator.py

```diff
@@ -439,6 +439,289 @@ def test_sop(self):
+        self.vocab_file = os.path.join(self.tmpdirname, "vocab.txt")
+        with open(self.vocab_file, "w", encoding="utf-8") as vocab_writer:
+            vocab_writer.write("".join([x + "\n" for x in vocab_tokens]))
+
+    def tearDown(self):
+        shutil.rmtree(self.tmpdirname)
+
+    def _turn_to_none(self, item):
+        """used to convert `item` to `None` type"""
+        return None
+
+    def _compare_assert_with_collator(self, collator, original, batch):
+        # we only care about side effects, the results are tested elsewhere
+        collator(batch)
+
+        # we go through every item and convert to `primitive` datatypes if necessary
+        # then compares for equivalence for the original data and the data that has been passed through the collator
+        for i in range(len(original)):
+            for key in original[i].keys():
+                if isinstance(original[i][key], np.ndarray):
+                    self.assertEqual(original[i][key].tolist(), batch[i][key].tolist())
+                elif isinstance(original[i][key], torch.Tensor):
+                    self.assertEqual(original[i][key].tolist(), batch[i][key].tolist())
+                else:
+                    self.assertEqual(original[i][key], batch[i][key])
+
+    def _compare_assert_with_collator_on_datatypes(
+        self, collator, base, input_key, input_datatype, label_key, label_datatype, ignore_label=False
+    ):
+        # using the arguments to recreate the features with their respective (potentially new) datatypes
+        features_original = [
+            {label_key: label_datatype(sample[label_key]), input_key: input_datatype(sample[input_key])}
+            for sample in base
+        ]
+        features_batch = [
+            {label_key: label_datatype(sample[label_key]), input_key: input_datatype(sample[input_key])}
+            for sample in base
+        ]
+
+        # some collators do not use labels, or sometimes we want to check if the collator with labels can handle such cases
+        if ignore_label:
+            for i in range(len(features_original)):
+                features_original[i].pop(label_key)
+                features_batch[i].pop(label_key)
+
+        self._compare_assert_with_collator(collator, features_original, features_batch)
+
+    def test_default_collator_immutability(self):
+        features_base_single_label = [{"label": i, "inputs": (0, 1, 2, 3, 4, 5)} for i in range(4)]
+        features_base_multiple_labels = [{"label": (0, 1, 2), "inputs": (0, 1, 2, 3, 4, 5)} for i in range(4)]
+
+        for datatype_input, datatype_label in [
+            (list, int),
+            (list, float),
+            (np.array, int),
+            (np.array, torch.tensor),
+            (list, self._turn_to_none),
+        ]:
+            self._compare_assert_with_collator_on_datatypes(
+                default_data_collator, features_base_single_label, "inputs", datatype_input, "label", datatype_label
```

Comment:
To make the test clearer - let's call the test with kwargs rather than args

Reply:
Propagated to other similar calls.